backend/backup_media.js

```diff
@@ -0,0 +1,155 @@
+  let dimensions = {width: 0, height: 0};
+  if (mediaType === MEDIA_TYPE_IMG) {
+    try {
+      dimensions = sizeOf(outputPath);
+    } catch (e) {}
+  }
+  let line = [
+    itemType, itemId, mediaType, mediaSrc, outputFilename, dimensions.width, dimensions.height
+  ];
+  fs.appendFileSync(config.dbFile, line.join(',') + "\n");
+}
+
+try { fs.mkdirSync(config.dir); } catch (e) {}
+if (!fs.existsSync(config.dbFile)) {
+  fs.openSync(config.dbFile, 'w+');
+  fs.appendFileSync(config.dbFile, 'item_type,item_id,media_type,media_src,media_id,img_width,img_height  ' + "\n");
+} else {
+  fs.openSync(config.dbFile, 'w+');
+}
+
+const connection = mysql.createConnection({
+  host: config.host,
+  user: config.user,
+  password: config.password,
+  database: config.database
+});
+
+const processBatch = async (offset) => {
+  const result = await connection.promise().query(
+    'select * from `content_source` where `source` like "%<img%" or `source` like "%<video%" limit ' + offset + ',' + BATCH_SIZE
+  );
+
+  if (!result[0] || !result[0].length) {
+    console.log('Done');
+    process.exit(0);
+  }
+
+  for (let i = 0; i < result[0].length; i++) {
+    let entry = result[0][i];
+    let itemType = entry.ref_type;
+    let itemId = entry.ref_id;
+    let entrySource = entry.source;
+    const mediaItems = entrySource.match(/<(img|source).+src=['"]([^'"]+)['"]/g);
+    if (!mediaItems) {
+      continue;
+    }
+    for (let ii = 0; ii < mediaItems.length; ii++) {
+      const check = mediaItems[ii].match(/<(img|source).+src=['"]([^'"]+)['"]/);
+      let mediaSrc = check[2];
+      let mediaType = check[1] === MEDIA_TYPE_IMG ? MEDIA_TYPE_IMG : MEDIA_TYPE_VIDEO;
+      let mediaExtensionMatch = mediaSrc.match(/\.[a-z]{2,}(?:\?.+)?$/);
+      let mediaExtension;
+      if (!mediaExtensionMatch) {
+        if (mediaType === MEDIA_TYPE_IMG) {
+          mediaExtension = '.jpg';
+        } else {
+          mediaExtension = '.mp4';
+        }
+      } else {
+        mediaExtension = mediaExtensionMatch[0].replace(/\?.+?$/, '');
+      }
+      let outputFilename = crypto.createHash('sha256').update(mediaSrc).digest('hex') + mediaExtension;
+      let outputPath = config.dir + '/' + outputFilename;
+      if (fs.existsSync(outputPath)) {
+        console.log(outputPath, 'is already there');
+        saveEntry(mediaType, outputPath, outputFilename, mediaSrc, itemType, itemId);
+        continue;
+      }
+      try {
+        console.log(entrySource);
+        console.log(`Downloading ${mediaSrc} to ${outputPath}`);
+        if (await download(mediaSrc, outputPath)) {
+          saveEntry(mediaType, outputPath, outputFilename, mediaSrc, itemType, itemId);
+          console.log('Downloaded: ', mediaSrc);
+        }
+      } catch (e) {
+        console.error('Failed to download: ', mediaSrc);
+      }
+    }
+  }
+  await sleep(SLEEP_BETWEEN_BATCHES_MS);
+  await processBatch(offset + BATCH_SIZE);
```

Comment:
I think it's more natural to return the number of processed items or a boolean "has next batch" flag from the `processBatch` method, and have the outer loop outside of the `processBatch` method.

Reply:
debatable, this will require saving total number into a var and comparing it with a number of processed entries which would need to be compared with the total

Reply:
debatable, this will require saving total number into a var and comparing it with a number of processed entries which would need to be compared with the total

Reply:
I mean, if `processBatch` returned `result[0].length` instead of calling `await processBatch(offset + BATCH_SIZE);` the "main" loop could look like this:

```js
let offset = 0
while (true) {
   const processed = await processBatch(offset);
   offset += processed
   if (processed < BATCH_SIZE) break;
}
```

The advantage of that is making `processBatch` have clearly defined responsibility: "process **batch**", instead of "process everything till the end".

Reply:
this makes sense, I'll adjust